File: src/main/java/me/felix/tired/modification/Module.java
package me.felix.tired.modification;

import lombok.Getter;
import lombok.Setter;
import me.felix.tired.api.Tired;
import me.felix.tired.bridge.event.Listener;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Getter
@Setter
public abstract class Module implements Listener {

    private final String name;
    private final ModuleCategory category;
    private int key;

    boolean toggled;

    public Module() {
        final Info info = getClass().getAnnotation(Info.class);
        this.name = info.name();
        this.category = info.category();
        this.key = info.defaultKey();
    }

    public boolean isToggled() {
        return toggled;
    }

    public void setToggled(boolean state) {
        this.toggled = state;
        if (state) {
            onEnable();
            Tired.addListener(this.getClass());
        } else {
            onDisable();
            Tired.removeListener(this.getClass());
        }
    }

    public abstract void onEnable();

    public abstract void onDisable();

    @Retention(RetentionPolicy.RUNTIME)
    @Target(ElementType.TYPE)
    public @interface Info {
        String name();

        ModuleCategory category();

        int defaultKey() default -1;
    }
}

File: src/main/kotlin/tired/jdk/api/module/Module.kt
package tired.jdk.api.module

import tired.jdk.api.Tired
import tired.jdk.api.event.Listener

abstract class Module : Listener {

    val name: String
    val category: Category
    var key: Int

    private var toggled = false

    init {
        val info = this.javaClass.getAnnotation(Info::class.java)
        this.name = info.name
        this.category = info.category
        this.key = info.defaultKey
    }

    fun isToggled() : Boolean {
        return toggled
    }

    fun setState(state: Boolean) {
        toggled = state
        if(state) {
            Tired.addListener(this.javaClass)
            onEnable()
        } else {
            Tired.removeListener(this.javaClass)
            onDisable()
        }
    }

    abstract fun onEnable()
    abstract fun onDisable()

    annotation class Info(val name: String, val category: Category, val defaultKey: Int = -1)

    enum class Category(val displayName: String) {
        COMBAT("Combat"), MOVEMENT("Movement"), VISUAL("Visual"), WORLD("World"), PLAYER("Player"), MISC("Misc")
    }
}